File: backend/app/pow.py
"""Client-side proof-of-work for the auth endpoints (bot cost-raising).

The server issues a signed challenge; the client must find a ``nonce`` such
that ``sha256(challenge + \".\" + nonce)`` has at least ``difficulty`` leading
zero bits (WebCrypto on the frontend — a few hundred milliseconds at 16
bits). The challenge is **stateless-signed** with ``SECRET_KEY`` (HMAC-SHA256
of ``challenge + issued_at``), so issuing needs no storage; verification
recomputes the signature and enforces the TTL.

Replay protection: on first successful verification the challenge is claimed
through the rate-limit store's ``consume_once`` primitive (single-use for the
TTL) — the same challenge cannot be replayed for a second request. If the
store is unavailable the claim fails open (the rate limiter still throttles).

``AUTH_POW_DIFFICULTY = 0`` disables the check entirely (dev / tests / proxy
environments) — the endpoint just accepts requests without a proof.
"""

from __future__ import annotations

import hashlib
import hmac
import secrets
import time

from .config import settings
from .ratelimit import consume_once

_SEP = "."


def _sign(challenge: str, issued_at: int) -> str:
    msg = f"{challenge}{_SEP}{issued_at}".encode()
    return hmac.new(settings.SECRET_KEY.encode(), msg, hashlib.sha256).hexdigest()


def issue_challenge() -> dict:
    """A fresh challenge for the client to solve.

    The client echoes ``challenge`` / ``issued_at`` / ``signature`` back with
    its ``nonce``; verification recomputes the signature from the key.
    """
    challenge = secrets.token_hex(16)
    issued_at = int(time.time())
    return {
        "challenge": challenge,
        "issued_at": issued_at,
        "signature": _sign(challenge, issued_at),
        "difficulty": settings.AUTH_POW_DIFFICULTY,
        "ttl_seconds": settings.AUTH_POW_TTL_SECONDS,
    }


def _leading_zero_bits(digest_hex: str) -> int:
    """Count leading zero bits of a hex-encoded 256-bit digest."""
    value = int(digest_hex, 16)
    return 256 - value.bit_length()


def verify(
    challenge: str,
    issued_at: int,
    signature: str,
    nonce: str,
) -> bool:
    """True when the proof is fresh, correctly signed and meets the difficulty.

    Consumes the challenge (single-use) on success — a replay of the same
    challenge returns False even with a valid signature + nonce.
    """
    if settings.AUTH_POW_DIFFICULTY <= 0:
        return True  # PoW disabled — nothing to verify
    if not challenge or not nonce:
        return False
    if not hmac.compare_digest(signature, _sign(challenge, issued_at)):
        return False
    age = int(time.time()) - issued_at
    if age < 0 or age > settings.AUTH_POW_TTL_SECONDS:
        return False
    digest = hashlib.sha256(f"{challenge}{_SEP}{nonce}".encode()).hexdigest()
    if _leading_zero_bits(digest) < settings.AUTH_POW_DIFFICULTY:
        return False
    # Claim the challenge so it can't be replayed within its TTL.
    return consume_once(f"pow:{challenge}", settings.AUTH_POW_TTL_SECONDS)

File: frontend/src/lib/pow.ts
/**
 * Client-side proof-of-work (hashcash) for the auth forms.
 *
 * The server issues a signed challenge (`GET /auth/pow-challenge`); the client
 * finds a `nonce` such that `SHA-256(challenge + "." + nonce)` has at least
 * `difficulty` leading zero bits. Solved asynchronously with WebCrypto so the
 * UI thread never blocks. The convention mirrors `app.pow` exactly — a wrong
 * separator or endianness here means every submit gets a 403.
 */

export interface PowChallenge {
  challenge: string
  issued_at: number
  signature: string
  difficulty: number
  ttl_seconds: number
}

export interface PowProof {
  challenge: string
  issued_at: number
  signature: string
  nonce: string
}

/** Leading zero bits of a 256-bit digest (big-endian byte order). */
function leadingZeroBits(bytes: Uint8Array): number {
  let zeros = 0
  for (const byte of bytes) {
    if (byte === 0) {
      zeros += 8
      continue
    }
    let v = byte
    while ((v & 0x80) === 0) {
      zeros += 1
      v <<= 1
    }
    break
  }
  return zeros
}

/**
 * Solve a challenge: return a nonce whose digest meets the difficulty.
 * Expected ~2^difficulty iterations; at 16 bits that is a few hundred ms.
 */
export async function solvePow(challenge: string, difficulty: number): Promise<string> {
  const encoder = new TextEncoder()
  const prefix = `${challenge}.`
  let nonce = 0
  for (;;) {
    const digest = await crypto.subtle.digest('SHA-256', encoder.encode(prefix + nonce))
    if (leadingZeroBits(new Uint8Array(digest)) >= difficulty) return String(nonce)
    nonce += 1
  }
}

/** Fetch a challenge and produce a proof (skips the work when disabled). */
export async function makePowProof(
  getChallenge: () => Promise<PowChallenge>,
): Promise<PowProof | null> {
  const challenge = await getChallenge()
  if (challenge.difficulty <= 0) return null
  const nonce = await solvePow(challenge.challenge, challenge.difficulty)
  return {
    challenge: challenge.challenge,
    issued_at: challenge.issued_at,
    signature: challenge.signature,
    nonce,
  }
}
